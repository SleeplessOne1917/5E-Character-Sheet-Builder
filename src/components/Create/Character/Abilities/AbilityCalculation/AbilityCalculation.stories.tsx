import AbilityCalculation, {
	AbilityCalculationProps
} from './AbilityCalculation';
import { ComponentMeta, ComponentStory } from '@storybook/react';

import MockStore from '../MockAbilitiesStore';

export default {
	title: 'Components/Create/Character/Abilities/AbilityCalculation',
	component: AbilityCalculation
} as ComponentMeta<typeof AbilityCalculation>;

const Template: ComponentStory<typeof AbilityCalculation> = (
	args: AbilityCalculationProps
) => <AbilityCalculation {...args} />;

export const OnlyBaseScore = Template.bind({});
OnlyBaseScore.args = {
	index: 'str',
	name: 'Strength'
};
OnlyBaseScore.decorators = [story => <MockStore>{story()}</MockStore>];

export const ScoreWithAbilityImprovement = Template.bind({});
ScoreWithAbilityImprovement.args = {
	index: 'dex',
	name: 'Dexterity'
};
ScoreWithAbilityImprovement.decorators = [
	story => <MockStore>{story()}</MockStore>
];

export const ScoreWithRaceBonus = Template.bind({});
ScoreWithRaceBonus.args = { index: 'cha', name: 'Charisma' };
ScoreWithRaceBonus.decorators = [story => <MockStore>{story()}</MockStore>];

export const OverrideWithNoBase = Template.bind({});
OverrideWithNoBase.args = {
	index: 'con',
	name: 'Constitution'
};
OverrideWithNoBase.decorators = [story => <MockStore>{story()}</MockStore>];

export const MiscBonusWithBase = Template.bind({});
MiscBonusWithBase.args = {
	index: 'wis',
	name: 'Wisdom'
};
MiscBonusWithBase.decorators = [
	story => (
		<MockStore
			overrideValues={{
				wis: { base: 13, miscBonus: 1, highest: 20, abilityImprovement: 0 }
			}}
		>
			{story()}
		</MockStore>
	)
];

export const OtherBonusWithNoBase = Template.bind({});
OtherBonusWithNoBase.args = {
	index: 'int',
	name: 'Intelligence'
};
OtherBonusWithNoBase.decorators = [story => <MockStore>{story()}</MockStore>];

export const BaseWithOverride = Template.bind({});
BaseWithOverride.args = {
	index: 'wis',
	name: 'Wisdom'
};
BaseWithOverride.decorators = [story => <MockStore>{story()}</MockStore>];

export const AllBonusesAndOverrideButNoBase = Template.bind({});
AllBonusesAndOverrideButNoBase.args = {
	index: 'dex',
	name: 'Dexterity'
};
AllBonusesAndOverrideButNoBase.decorators = [
	story => (
		<MockStore
			overrideValues={{
				dex: {
					base: null,
					abilityImprovement: 4,
					raceBonus: 2,
					otherBonus: 1,
					override: 28,
					miscBonus: 1,
					highest: 20
				}
			}}
		>
			{story()}
		</MockStore>
	)
];

export const Everything = Template.bind({});
Everything.args = {
	index: 'str',
	name: 'Strength'
};
Everything.decorators = [
	story => (
		<MockStore
			overrideValues={{
				str: {
					base: 15,
					abilityImprovement: 2,
					miscBonus: 3,
					raceBonus: 1,
					otherBonus: 1,
					override: 5,
					highest: 20
				}
			}}
		>
			{story()}
		</MockStore>
	)
];
