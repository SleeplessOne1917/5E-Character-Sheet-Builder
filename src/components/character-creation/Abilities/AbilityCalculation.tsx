import {
	ChangeEventHandler,
	FocusEventHandler,
	useCallback,
	useState
} from 'react';
import {
	calculateModifier,
	getTotalScore
} from '../../../services/abilityScoreService';
import {
	updateOtherBonus,
	updateOverride
} from '../../../redux/features/abilityScores';

import AbilityScores from '../../../types/abilityScores';
import classes from './AbilityCalculation.module.css';
import { useAppDispatch } from '../../../hooks/reduxHooks';
import useGetAbilityScore from '../../../hooks/useGetAbilityScore';

type AbilityCalculationProps = {
	index: string;
	name: string;
};

const AbilityCalculation = ({ index, name }: AbilityCalculationProps) => {
	const [otherBonus, setOtherBonus] = useState<number | undefined>(undefined);
	const [overrideScore, setOverrideScore] = useState<number | undefined>(
		undefined
	);
	const dispatch = useAppDispatch();
	const getAbilityScore = useGetAbilityScore();

	const onChangeOtherBonus: ChangeEventHandler<HTMLInputElement> = useCallback(
		event => {
			setOtherBonus(parseInt(event.target.value));
		},
		[setOtherBonus]
	);

	const onBlurOtherBonus: FocusEventHandler<HTMLInputElement> = useCallback(
		event => {
			let value: number | null = parseInt(event.target.value);

			if (isNaN(value)) {
				value = null;
			} else {
				if (value < -10) {
					value = -10;
				}

				if (value > 10) {
					value = 10;
				}
			}

			setOtherBonus(value ? value : undefined);
			dispatch(
				updateOtherBonus({ value, abilityIndex: index as AbilityScores })
			);
		},
		[dispatch, index, setOtherBonus]
	);

	const onChangeOverrideScore: ChangeEventHandler<HTMLInputElement> =
		useCallback(
			event => {
				setOverrideScore(parseInt(event.target.value));
			},
			[setOverrideScore]
		);

	const onBlurOverrideScore: FocusEventHandler<HTMLInputElement> = useCallback(
		event => {
			let value: number | null = parseInt(event.target.value);

			if (isNaN(value)) {
				value = null;
			} else {
				if (value < 3) {
					value = 3;
				}

				if (value > 30) {
					value = 30;
				}
			}

			setOverrideScore(value ? value : undefined);
			dispatch(updateOverride({ value, abilityIndex: index as AbilityScores }));
		},
		[dispatch, index, setOverrideScore]
	);

	const abilityScore = getAbilityScore(index as AbilityScores);
	const modifier = abilityScore.base
		? calculateModifier(getTotalScore(abilityScore))
		: null;
	return (
		<div className={classes.calculation}>
			<div className={classes.header}>
				<svg className={classes.icon}>
					<use xlinkHref={`/Icons.svg#${index}`} />
				</svg>
				{name}
			</div>
			<div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Total Score
					</div>
					<div className={classes.value}>
						{abilityScore.base ? getTotalScore(abilityScore) : '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Modifier
					</div>
					<div className={classes.value}>
						{modifier !== null
							? `${modifier < 0 ? '-' : '+'}${Math.abs(modifier)}`
							: '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Base Score
					</div>
					<div className={classes.value}>
						{abilityScore.base ? abilityScore.base : '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Racial Bonus
					</div>
					<div className={classes.value}>
						+{abilityScore.raceBonus ? abilityScore.raceBonus : 0}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Ability Improvements
					</div>
					<div className={classes.value}>
						+
						{abilityScore.abilityImprovement
							? abilityScore.abilityImprovement
							: 0}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Misc Bonus
					</div>
					<div className={classes.value}>
						+{abilityScore.miscBonus ? abilityScore.miscBonus : 0}
					</div>
				</div>
				<div className={classes.component}>
					<div className={classes.label}>
						<label htmlFor={`other-${index}`}>Other Modifier</label>
					</div>
					<div className={classes.value}>
						<input
							id={`other-${index}`}
							type="number"
							placeholder="&mdash;"
							value={otherBonus}
							onChange={onChangeOtherBonus}
							onBlur={onBlurOtherBonus}
						/>
					</div>
				</div>
				<div className={classes.component}>
					<div className={classes.label}>
						<label htmlFor={`override-${index}`}>Override Score</label>
					</div>
					<div className={classes.value}>
						<input
							id={`override-${index}`}
							type="number"
							placeholder="&mdash;"
							value={overrideScore}
							onChange={onChangeOverrideScore}
							onBlur={onBlurOverrideScore}
						/>
					</div>
				</div>
			</div>
		</div>
	);
};

export default AbilityCalculation;
