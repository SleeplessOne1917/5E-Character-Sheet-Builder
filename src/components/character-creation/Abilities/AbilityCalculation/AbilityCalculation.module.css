.calculation {
    width: 100%
}

.icon {
    height: 1.5rem;
    width: 1.5rem;
    fill: var(--heaven);
    margin-right: 0.7rem;
}

.header {
    background-color: var(--davy);
    color: var(--heaven);
    font-family: var(--font-fantasy);
    padding: 0.5rem;
    font-size: 1.2rem;
    display: flex;
    align-items: center;
}

.component {
    display: flex;
    font-size: 1.2rem;
    background-color: var(--heaven);
}

.calculation-component {
    border: 1px solid var(--davy);
}

.label {
    width: 70%;
    padding: 0.5rem;
    background-color: var(--heaven);
    display: flex;
    align-items: center;
}

.calculation-label {
    border-right: 2px solid var(--highlight);
}

.value {
    width: 30%;
    padding: 0.5rem;
    text-align: center;
    font-size: 1.5rem;
    background-color: var(--heaven-light);
}

.input-value {
    background-color: var(--heaven) !important;
}

.value input {
    width: 100%;
    font-size: 1.5rem;
    font-family: var(--font-display);
    text-align: center;
    border: 1px solid var(--davy);
    background-color: var(--heaven-light);
}

.value input:focus {
    outline: var(--focus-outline);
}

@media screen and (min-width: 768px) {
    .calculation {
        width: 30%;
        margin: 0.6rem;
    }

    .component {
        font-size: 0.85rem;
    }
}

@media screen and (min-width: 1280px) {
    .component {
        font-size: 1rem;
    }
}