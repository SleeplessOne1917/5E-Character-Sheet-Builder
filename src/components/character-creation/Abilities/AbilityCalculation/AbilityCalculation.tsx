import {
	ChangeEventHandler,
	FocusEventHandler,
	useCallback,
	useState
} from 'react';
import {
	calculateModifier,
	getTotalScore
} from '../../../../services/abilityScoreService';
import {
	updateOtherBonus,
	updateOverride
} from '../../../../redux/features/abilityScores';

import AbilityScores from '../../../../types/abilityScores';
import classes from './AbilityCalculation.module.css';
import { useAppDispatch } from '../../../../hooks/reduxHooks';
import useGetAbilityScore from '../../../../hooks/useGetAbilityScore';

export type AbilityCalculationProps = {
	index: string;
	name: string;
};

const AbilityCalculation = ({
	index,
	name
}: AbilityCalculationProps): JSX.Element => {
	const getAbilityScore = useGetAbilityScore();
	const abilityScore = getAbilityScore(index as AbilityScores);
	const [otherBonus, setOtherBonus] = useState<number | undefined>(
		abilityScore.otherBonus || undefined
	);
	const [overrideScore, setOverrideScore] = useState<number | undefined>(
		abilityScore.override || undefined
	);
	const dispatch = useAppDispatch();

	const onChangeOtherBonus: ChangeEventHandler<HTMLInputElement> = useCallback(
		event => {
			setOtherBonus(parseInt(event.target.value));
		},
		[setOtherBonus]
	);

	const onBlurOtherBonus: FocusEventHandler<HTMLInputElement> = useCallback(
		event => {
			let value: number | null = parseInt(event.target.value);

			if (isNaN(value)) {
				value = null;
			} else {
				if (value < -10) {
					value = -10;
				}

				if (value > 10) {
					value = 10;
				}
			}

			setOtherBonus(value ? value : undefined);
			dispatch(
				updateOtherBonus({ value, abilityIndex: index as AbilityScores })
			);
		},
		[dispatch, index, setOtherBonus]
	);

	const onChangeOverrideScore: ChangeEventHandler<HTMLInputElement> =
		useCallback(
			event => {
				setOverrideScore(parseInt(event.target.value));
			},
			[setOverrideScore]
		);

	const onBlurOverrideScore: FocusEventHandler<HTMLInputElement> = useCallback(
		event => {
			let value: number | null = parseInt(event.target.value);

			if (isNaN(value)) {
				value = null;
			} else {
				if (value < 3) {
					value = 3;
				}

				if (value > 30) {
					value = 30;
				}
			}

			setOverrideScore(value ? value : undefined);
			dispatch(updateOverride({ value, abilityIndex: index as AbilityScores }));
		},
		[dispatch, index, setOverrideScore]
	);

	const modifier = abilityScore.base
		? calculateModifier(getTotalScore(abilityScore))
		: null;
	return (
		<div className={classes.calculation}>
			<div className={classes.header}>
				<svg className={classes.icon}>
					<use xlinkHref={`/Icons.svg#${index}`} />
				</svg>
				{name}
			</div>
			<div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Total Score
					</div>
					<div className={classes.value}>
						{abilityScore.base ? getTotalScore(abilityScore) : '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Modifier
					</div>
					<div className={classes.value}>
						{modifier !== null
							? `${modifier < 0 ? '-' : '+'}${Math.abs(modifier)}`
							: '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Base Score
					</div>
					<div className={classes.value}>
						{abilityScore.base ? abilityScore.base : '\u2014'}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Racial Bonus
					</div>
					<div className={classes.value}>
						+{abilityScore.raceBonus ? abilityScore.raceBonus : 0}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Ability Improvements
					</div>
					<div className={classes.value}>
						+
						{abilityScore.abilityImprovement
							? abilityScore.abilityImprovement
							: 0}
					</div>
				</div>
				<div
					className={`${classes.component} ${classes['calculation-component']}`}
				>
					<div className={`${classes.label} ${classes['calculation-label']}`}>
						Misc Bonus
					</div>
					<div className={classes.value}>
						+{abilityScore.miscBonus ? abilityScore.miscBonus : 0}
					</div>
				</div>
				<div className={classes.component}>
					<div className={classes.label}>
						<label htmlFor={`other-${index}`}>Other Bonus</label>
					</div>
					<div className={`${classes.value} ${classes['input-value']}`}>
						<input
							id={`other-${index}`}
							type="number"
							placeholder="&mdash;"
							value={otherBonus}
							onChange={onChangeOtherBonus}
							onBlur={onBlurOtherBonus}
						/>
					</div>
				</div>
				<div className={classes.component}>
					<div className={classes.label}>
						<label htmlFor={`override-${index}`}>Override Score</label>
					</div>
					<div className={`${classes.value} ${classes['input-value']}`}>
						<input
							id={`override-${index}`}
							type="number"
							placeholder="&mdash;"
							value={overrideScore}
							onChange={onChangeOverrideScore}
							onBlur={onBlurOverrideScore}
						/>
					</div>
				</div>
			</div>
		</div>
	);
};

export default AbilityCalculation;
