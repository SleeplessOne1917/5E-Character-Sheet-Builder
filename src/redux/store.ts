import {
	AnyAction,
	ConfigureStoreOptions,
	ListenerEffect,
	Reducer,
	ThunkDispatch,
	configureStore,
	createListenerMiddleware
} from '@reduxjs/toolkit';
import editingCharacter, {
	EditingCharacterState
} from './features/editingCharacter';
import { get, set } from 'idb-keyval';
import { useEffect, useState } from 'react';

import generationMethod from './features/generationMethod';
import rollGroups from './features/rollGroups';
import toast from './features/toast';
import viewer from './features/viewer';

const getReducer = () => ({
	toast,
	viewer,
	editingCharacter: editingCharacter as Reducer<
		EditingCharacterState,
		AnyAction
	>,
	rollGroups,
	generationMethod
});

const indexedDbCacheListener = createListenerMiddleware();

const REDUX_CACHE_KEY = 'redux-cache';

const updateCache: ListenerEffect<
	AnyAction,
	unknown,
	ThunkDispatch<unknown, unknown, AnyAction>,
	unknown
> = (action, { getState }) => {
	set(REDUX_CACHE_KEY, getState());
};

const updateCachePredicate = () => true;

const getStore = async () => {
	indexedDbCacheListener.startListening({
		predicate: updateCachePredicate,
		effect: updateCache
	});

	const options: ConfigureStoreOptions = {
		reducer: getReducer(),
		middleware: getDefaultMiddleware =>
			getDefaultMiddleware().prepend(indexedDbCacheListener.middleware)
	};

	const reduxCache = await get(REDUX_CACHE_KEY);
	if (reduxCache) {
		options.preloadedState = reduxCache;
	}

	const theStore = configureStore(options);
	if (!reduxCache) {
		set(REDUX_CACHE_KEY, theStore.getState());
	}

	return theStore;
};

export const useStore = () => {
	const [theStore, setTheStore] = useState(getTestStore());
	const [loadingStore, setLoadingStore] = useState(true);

	useEffect(() => {
		getStore().then(s => {
			setTheStore(s);
			setLoadingStore(false);
		});

		return () => {
			indexedDbCacheListener.stopListening({
				predicate: updateCachePredicate,
				effect: updateCache
			});
		};
	}, [setTheStore]);

	return { store: theStore, loading: loadingStore };
};

export const getTestStore = (
	reducerOverrides?: Partial<ReturnType<typeof getReducer>>
) =>
	configureStore({
		reducer: { ...getReducer(), ...reducerOverrides },
		devTools: false
	});

const store = getTestStore();

export type RootState = ReturnType<typeof store.getState>;
export type Dispatch = typeof store.dispatch;
