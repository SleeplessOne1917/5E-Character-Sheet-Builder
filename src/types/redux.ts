import { SliceCaseReducers, ValidateSliceCaseReducers } from '@reduxjs/toolkit';

export type AppReducers<T> = ValidateSliceCaseReducers<T, SliceCaseReducers<T>>;
