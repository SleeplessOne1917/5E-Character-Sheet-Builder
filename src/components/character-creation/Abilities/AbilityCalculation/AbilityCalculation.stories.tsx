import AbilityCalculation, {
	AbilityCalculationProps
} from './AbilityCalculation';
import {
	AbilityScore,
	AbilityScoresState
} from '../../../../redux/features/abilityScores';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { configureStore, createSlice } from '@reduxjs/toolkit';

import AbilityScores from '../../../../types/abilityScores';
import { Provider } from 'react-redux';
import { ReactNode } from 'react';

const defaultInitialState: { abilityScores: AbilityScoresState } = {
	abilityScores: {
		str: {
			base: 12
		},
		dex: {
			base: 8,
			abilityImprovement: 4
		},
		cha: { base: 15, raceBonus: 2 },
		con: { override: 10 },
		int: { otherBonus: 5 },
		wis: { base: 14, override: 25 }
	}
};

const getMockStore = (index: AbilityScores, overrideValues: AbilityScore) => {
	const initialState = {
		abilityScores: {
			...defaultInitialState.abilityScores,
			[index]: {
				...defaultInitialState.abilityScores[index],
				...overrideValues
			}
		}
	};

	const mockEditingCharacterSlice = createSlice({
		name: 'editingCharacter',
		initialState: initialState,
		reducers: {}
	});

	return configureStore({
		reducer: { editingCharacter: mockEditingCharacterSlice.reducer }
	});
};

const MockStore = ({
	overrideValues = {},
	index,
	children
}: {
	overrideValues?: AbilityScore;
	index: AbilityScores;
	children: ReactNode;
}) => (
	<Provider store={getMockStore(index, overrideValues)}>{children}</Provider>
);

const AbilityCalculationSb = {
	title: 'Components/AbilityCalculation',
	component: AbilityCalculation
} as ComponentMeta<typeof AbilityCalculation>;

export default AbilityCalculationSb;

const Template: ComponentStory<typeof AbilityCalculation> = (
	args: AbilityCalculationProps
) => <AbilityCalculation {...args} />;

export const OnlyBaseScore = Template.bind({});
OnlyBaseScore.args = {
	index: 'str',
	name: 'Strength'
};
OnlyBaseScore.decorators = [
	story => <MockStore index="str">{story()}</MockStore>
];

export const ScoreWithAbilityImprovement = Template.bind({});
ScoreWithAbilityImprovement.args = {
	index: 'dex',
	name: 'Dexterity'
};
ScoreWithAbilityImprovement.decorators = [
	story => <MockStore index="dex">{story()}</MockStore>
];

export const ScoreWithRaceBonus = Template.bind({});
ScoreWithRaceBonus.args = { index: 'cha', name: 'Charisma' };
ScoreWithRaceBonus.decorators = [
	story => <MockStore index="cha">{story()}</MockStore>
];

export const OverrideWithNoBase = Template.bind({});
OverrideWithNoBase.args = {
	index: 'con',
	name: 'Constitution'
};
OverrideWithNoBase.decorators = [
	story => <MockStore index="con">{story()}</MockStore>
];

export const MiscBonusWithBase = Template.bind({});
MiscBonusWithBase.args = {
	index: 'wis',
	name: 'Wisdom'
};
MiscBonusWithBase.decorators = [
	story => (
		<MockStore index="wis" overrideValues={{ base: 13, miscBonus: 1 }}>
			{story()}
		</MockStore>
	)
];

export const OtherBonusWithNoBase = Template.bind({});
OtherBonusWithNoBase.args = {
	index: 'int',
	name: 'Intelligence'
};
OtherBonusWithNoBase.decorators = [
	story => <MockStore index="int">{story()}</MockStore>
];

export const BaseWithOverride = Template.bind({});
BaseWithOverride.args = {
	index: 'wis',
	name: 'Wisdom'
};
BaseWithOverride.decorators = [
	story => <MockStore index="wis">{story()}</MockStore>
];

export const AllBonusesAndOverrideButNoBase = Template.bind({});
AllBonusesAndOverrideButNoBase.args = {
	index: 'dex',
	name: 'Dexterity'
};
AllBonusesAndOverrideButNoBase.decorators = [
	story => (
		<MockStore
			index="dex"
			overrideValues={{
				base: null,
				abilityImprovement: 4,
				raceBonus: 2,
				otherBonus: 1,
				override: 28,
				miscBonus: 1
			}}
		>
			{story()}
		</MockStore>
	)
];

export const Everything = Template.bind({});
Everything.args = {
	index: 'str',
	name: 'Strength'
};
Everything.decorators = [
	story => (
		<MockStore
			index="str"
			overrideValues={{
				base: 15,
				abilityImprovement: 2,
				miscBonus: 3,
				raceBonus: 1,
				otherBonus: 1,
				override: 5
			}}
		>
			{story()}
		</MockStore>
	)
];
