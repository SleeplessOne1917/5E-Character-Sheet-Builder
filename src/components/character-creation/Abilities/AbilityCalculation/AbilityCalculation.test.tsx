import '@testing-library/jest-dom';

import * as stories from './AbilityCalculation.stories';

import { render, screen } from '@testing-library/react';

import AbilityCalculation from './AbilityCalculation';
import { Provider } from 'react-redux';
import { composeStories } from '@storybook/testing-react';
import { getTestStore } from '../../../../redux/store';
import userEvent from '@testing-library/user-event';

const { AllBonusesAndOverrideButNoBase, Everything, OnlyBaseScore } =
	composeStories(stories);

it('renders correctly', () => {
	render(<Everything />);

	expect(screen.getByTestId('ability-calculation')).toMatchSnapshot();
});

it('displays total score and modifier when base score is set', () => {
	render(<OnlyBaseScore />);

	expect(screen.getByTestId('total-score')).not.toHaveTextContent('\u2014');
	expect(screen.getByTestId('modifier')).not.toHaveTextContent('\u2014');
});

it('does not display total score when base score is not set', () => {
	render(<AllBonusesAndOverrideButNoBase />);

	expect(screen.getByTestId('total-score')).toHaveTextContent('\u2014');
	expect(screen.getByTestId('modifier')).toHaveTextContent('\u2014');
	expect(screen.getByTestId('base-score')).toHaveTextContent('\u2014');
});

it('uses default values when bonuses not passed in', () => {
	render(<OnlyBaseScore />);

	expect(screen.getByTestId('racial-bonus')).toHaveTextContent('+0');
	expect(screen.getByTestId('ability-improvements')).toHaveTextContent('+0');
	expect(screen.getByTestId('misc-bonus')).toHaveTextContent('+0');
});

describe('other bonus', () => {
	it('does not allow input lower than -10', async () => {
		render(
			<Provider store={getTestStore()}>
				<AbilityCalculation index="str" name="Strength" />
			</Provider>
		);
		const otherBonusInput = screen.getByLabelText(
			/Other Bonus/i
		) as HTMLInputElement;
		await userEvent.type(otherBonusInput, '-11');
		await userEvent.click(screen.getByText(/Other Bonus/i));

		expect(otherBonusInput.value).toBe('-10');
	});

	it('does not allow input greater than 10', async () => {
		render(
			<Provider store={getTestStore()}>
				<AbilityCalculation index="dex" name="Dexterity" />
			</Provider>
		);
		const otherBonusInput = screen.getByLabelText(
			/Other Bonus/i
		) as HTMLInputElement;
		await userEvent.type(otherBonusInput, '11');
		await userEvent.click(screen.getByText(/Other Bonus/i));

		expect(otherBonusInput.value).toBe('10');
	});
});

describe('override', () => {
	it('does not allow input lower than 3', async () => {
		render(
			<Provider store={getTestStore()}>
				<AbilityCalculation index="con" name="Constitution" />
			</Provider>
		);
		const overrideBonusInput = screen.getByLabelText(
			/Override Score/i
		) as HTMLInputElement;
		await userEvent.type(overrideBonusInput, '2');
		await userEvent.click(screen.getByText(/Override Score/i));

		expect(overrideBonusInput.value).toBe('3');
	});

	it('does not allow input greater than 30', async () => {
		render(
			<Provider store={getTestStore()}>
				<AbilityCalculation index="con" name="Constitution" />
			</Provider>
		);
		const overrideBonusInput = screen.getByLabelText(
			/Override Score/i
		) as HTMLInputElement;
		await userEvent.type(overrideBonusInput, '31');
		await userEvent.click(screen.getByText(/Override Score/i));

		expect(overrideBonusInput.value).toBe('30');
	});
});
