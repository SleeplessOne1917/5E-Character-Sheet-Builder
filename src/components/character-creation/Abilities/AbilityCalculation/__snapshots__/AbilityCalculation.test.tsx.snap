// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="calculation"
  data-testid="ability-calculation"
>
  <div
    class="header"
  >
    <svg
      class="icon"
    >
      <use
        xlink:href="/Icons.svg#str"
      />
    </svg>
    Strength
  </div>
  <div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Total Score
      </div>
      <div
        class="value"
        data-testid="total-score"
      >
        5
      </div>
    </div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Modifier
      </div>
      <div
        class="value"
        data-testid="modifier"
      >
        -3
      </div>
    </div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Base Score
      </div>
      <div
        class="value"
        data-testid="base-score"
      >
        15
      </div>
    </div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Racial Bonus
      </div>
      <div
        class="value"
        data-testid="racial-bonus"
      >
        +
        1
      </div>
    </div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Ability Improvements
      </div>
      <div
        class="value"
        data-testid="ability-improvements"
      >
        +
        2
      </div>
    </div>
    <div
      class="component calculation-component"
    >
      <div
        class="label calculation-label"
      >
        Misc Bonus
      </div>
      <div
        class="value"
        data-testid="misc-bonus"
      >
        +
        3
      </div>
    </div>
    <div
      class="component"
    >
      <div
        class="label"
      >
        <label
          for="other-str"
        >
          Other Bonus
        </label>
      </div>
      <div
        class="value input-value"
      >
        <input
          id="other-str"
          placeholder="—"
          type="number"
          value="1"
        />
      </div>
    </div>
    <div
      class="component"
    >
      <div
        class="label"
      >
        <label
          for="override-str"
        >
          Override Score
        </label>
      </div>
      <div
        class="value input-value"
      >
        <input
          id="override-str"
          placeholder="—"
          type="number"
          value="5"
        />
      </div>
    </div>
  </div>
</div>
`;
